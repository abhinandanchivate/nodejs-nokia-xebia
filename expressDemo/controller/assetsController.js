//every endpoint from the assets ==> checking it ?
// we will write it at one place and we will apply it at multiple one .

import { validationResult } from "express-validator";
import AssetsModel from "../models/AssetsModel.js";
import UserModel from "../models/UserModel.js";

export const createAsset = async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }
  try {
    const asset = new AssetsModel(req.body);
    // assignedto email id exists or not.==> Users
    const user = await UserModel.find({ email: req.body.assignedTo });
    console.log("user value " + user);
    if (!user || user.length === 0) {
      console.log("inside the user criteria");
      // return res.status(404).json({ message: "User not found" });
      throw new Error("user not found");
    } else {
      await asset.save();
      res.status(201).json(asset);
    }
  } catch (error) {
    console.log("inside the catch ");
    throw new Error(error.message);
    //res.status(500).json({ error: error.message });
  }
};
export const getAllAsset = async (req, res) => {
  try {
    const assets = await AssetsModel.find();
    if (assets.length === 0) {
      return res.status(404).json({ message: "No assets found" });
    }
    res.status(200).json(assets);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
};
export const getAssetById = async (req, res) => {
  // whatever the id is provided 1st print that id
  const id = req.params.id;

  //id = 6853f02676f13fb3aa2d1d75
  const asset = await AssetsModel.findById(id);
  if (!asset) {
    return res.status(404).json({ message: "Asset not found" });
  } else {
    res.status(200).json(asset);
  }
};
export const updateAsset = async (req, res) => {};
export const deleteAsset = async (req, res) => {};
