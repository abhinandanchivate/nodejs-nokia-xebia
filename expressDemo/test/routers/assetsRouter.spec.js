// // test/routers/assetsRouter.spec.js
// import request from "supertest";
// import express from "express";
// import * as controller from "../../controller/assetsController.js";
// import { describe, it, beforeEach, vi, expect } from "vitest";

// // Dynamically import router *after* mocking
// let app;

// beforeEach(async () => {
//   vi.restoreAllMocks(); // clear previous spies

//   // Mock controller before importing router
//   vi.spyOn(controller, "getAllAsset").mockImplementation((req, res) => {
//     return res.status(200).json([
//       { id: 1, name: "Laptop", assignedTo: "john@example.com" },
//       { id: 2, name: "Monitor", assignedTo: "jane@example.com" },
//     ]);
//   });

//   // Recreate app with freshly imported router (now using the mock)
//   const { default: assetsRouter } = await import(
//     "../../router/assetsRouter.js"
//   );

//   app = express();
//   app.use(express.json());
//   app.use("/api/v1/assets", assetsRouter);
// });

// it("GET /api/v1/assets/all - should return snapshot", async () => {
//   const response = await request(app).get("/api/v1/assets/all");
//   expect(response.status).toBe(200);
//   console.log(response.body);
//   expect(response.body).toMatchSnapshot();
// }, 10000);

import request from "supertest";
import express from "express";
import * as controller from "../../controller/assetsController.js";
import { describe, it, beforeEach, vi, expect } from "vitest";

// ✅ Stub auth middleware to always allow
vi.mock("../../middleware/authMiddleware.js", () => ({
  authMiddleware: (req, res, next) => next(),
}));

let app;

beforeEach(async () => {
  vi.restoreAllMocks(); // clear previous spies

  // ✅ Mock getAssetById before loading router
  vi.spyOn(controller, "getAssetById").mockImplementation((req, res) => {
    return res.status(200).json({
      id: req.params.id,
      name: "Keyboard",
      assignedTo: "mark@example.com",
    });
  });

  const { default: assetsRouter } = await import(
    "../../router/assetsRouter.js"
  );

  app = express();
  app.use(express.json());
  app.use("/api/v1/assets", assetsRouter);
});

it("GET /api/v1/assets/:id - should return single asset snapshot", async () => {
  const res = await request(app).get("/api/v1/assets/123");

  expect(res.status).toBe(200);
  expect(res.body).toMatchSnapshot();
});
