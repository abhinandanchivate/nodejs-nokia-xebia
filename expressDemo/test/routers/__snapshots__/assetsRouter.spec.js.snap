// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /api/v1/assets/:id - should return single asset snapshot 1`] = `
{
  "assignedTo": "mark@example.com",
  "id": "123",
  "name": "Keyboard",
}
`;

exports[`GET /api/v1/assets/all - should return snapshot 1`] = `
[
  {
    "assignedTo": "john@example.com",
    "id": 1,
    "name": "Laptop",
  },
  {
    "assignedTo": "jane@example.com",
    "id": 2,
    "name": "Monitor",
  },
]
`;
